### ISSUE TITLE
Enable GCC and Clang warnings about nonstandard `printf` format strings

### ISSUE THREAD

**Issue by astiob** [libass/libass#529]
GCC needs `-Wpedantic`, which also displays a few other warnings that we may want to fix. Clang also supports this flag and displays a few other warnings (mostly the same as GCC), but nonstandard format string warnings are not among them.

Clang needs `-Wformat-non-iso`, which GCC doesn’t support (it exits with an error about an unrecognized flag).

We need a `configure` check for `-Wformat-non-iso` at the very least. We probably want to do configure checks for both of these and also for those options that we currently unconditionally add in `Makefile.am`.

**Comment by rcombs** [comment]
I don't think `-Wpedantic` is a great call here; it's extremely picky about plenty of stuff that's entirely well-defined and sensible on all supported platforms.

What specific cases are we actually worried about here? Using C11 format string modifiers while building for GNU99? I think the more appropriate route at this point is to just bump the build to GNU11 and note that we require a sufficiently recent (i.e. ten years!) compiler and standard library.

**Comment by astiob** [comment]
> Using C11 format string modifiers while building for GNU99?

No, just using formats that aren’t standard C at all, like `%m`. And the only way to get warned about `%m` is to use GCC and give it `-Wpedantic`.

(FWIW I don’t think standard `printf` formats have been extended since C99.)

> it's extremely picky about plenty of stuff that's entirely well-defined and sensible on all supported platforms

You’re not wrong. But then again, it gives only a few warnings for current libass, and they’re easy to fix, even if the fixes do introduce a little noise. And it’s the only way to get nonstandard-format warnings.

> bump the build to GNU11 and note that we require a sufficiently recent (i.e. ten years!) compiler and standard library.

For that matter, it’s not just about the standard library. **Our log messages do not go to the standard library. They are part of our own API.** They’re parsed by the log callback, which is implemented by the consumer, so whenever we start using any new format specifier, we’re breaking the API in a way that can’t even be automatically detected at compile-time.

Even now, we neither explicitly document what formats we use nor are consistent about it: we use `%zu`, and at the same time we use `PRId64` (which may be `%I64d`) with `(int64_t) long_long_value`.

It seems we have a majority consensus that using standard C99 formats is OK, but I’d prefer that we document this, stick to it consistently, and then avoid any future standard formats until/unless we really need them, even if we eagerly start using `-std=gnu23+` to compile ourselves.

**Comment by rcombs** [comment]
Are we worried about someone using `%m` accidentally? As far as I can tell, this has never been an issue. I wouldn't object to adding a warning just for this, but I don't think it's worth going as broad as `-Wpedantic` (which wouldn't even help with cases where a flag _is_ in the standard but isn't handled by some nonconformant libc we may or may not care about, like older MSVCRT).

> (FWIW I don’t think standard printf formats have been extended since C99.)

Ah, yeah I'd misread; indeed these aren't new.

> Our log messages do not go to the standard library. They are part of our own API.

In practice, they always go to the standard library (I would be _shocked_ if any caller parsed them themselves). These `const char*, va_list` callbacks are common and the expectation is that the caller will pass those args to one of the `vprintf` family of functions (usually `vsprintf`). We can document that if we want to be extra-clear, but I don't see any reason to limit ourselves to anything less than "whatever the libc supports". The `PRI[…]` macros using a particular modifier implies that the `v[sn]printf` implementation will support it, and the only C99 thing I'd be at all concerned about is `%zu` (and I have no qualms about requiring a libc that supports that, since MSVCRT has handled it fine for years).

**Comment by astiob** [comment]
> `%zu` (and I have no qualms about requiring a libc that supports that, since MSVCRT has handled it fine for years).

MSVCRT does *not* support `%zu`.

**Comment by rcombs** [comment]
It does since at least VS2015, and is described in the microsoft docs for that version: https://docs.microsoft.com/en-us/cpp/c-runtime-library/format-specification-syntax-printf-and-wprintf-functions?view=msvc-140#size-prefixes-for-printf-and-wprintf-format-type-specifiers

**Comment by astiob** [comment]
VS2015+ uses UCRT. And VS has not used MSVCRT since 2004. MinGW-w64 and MinGW.org use MSVCRT by default, but they are able to provide their own `printf` as an inline function in `stdio.h`.

---

For some extended context that this issue came from, see https://github.com/libass/libass/pull/523#discussion_r664082141 and d7e0e17ce571a8eca8fdf978a7c00c096782d037.

**Comment by rcombs** [comment]
Ahhh, sorry, I hadn't realized they'd changed the name. To rephrase: windows has shipped a working printf implementation for years, so I have no qualms about requiring(/continuing to require) `%zu` support.

**Comment by astiob** [comment]
Right. Well, technically, libass’s consumer can still be built with MinGW in C90 mode or so and end up calling actual-MSVCRT `printf` that still doesn’t support `%zu`. But like I tried to say—it seems the majority consensus is that this case isn’t important enough for us to care.

> Are we worried about someone using `%m` accidentally?

Yep. Or whatever other format extension glibc and GCC may support. Right, @TheOneric?

Based on this discussion, I suppose your original intention for the `format` attribute was merely to catch cases when the format string is valid-enough but the arguments don’t match?

> As far as I can tell, this has never been an issue.

Not yet :-)

> I wouldn't object to adding a warning just for this, but I don't think it's worth going as broad as `-Wpedantic`

And you may well be right.

We _may_ settle on `-Wformat-non-iso` for Clang and nothing for GCC, although that still requires a configure check. Unfortunately, it doesn’t catch `%m` specifically, but I assume there are other things that it does catch.

> (which wouldn't even help with cases where a flag _is_ in the standard but isn't handled by some nonconformant libc we may or may not care about, like older MSVCRT)

That’s a bit of a separate (though certainly related) problem. Ultimately, we’re trying to tackle that by using `format(printf)` in GCC on non-Windows, which theoretically (some day in the future, or on some platform that adds a custom patch to it) might be more attuned to the system libc than `format(gnu_printf)` is. And also by deciding that we don’t officially support such platforms.

**Comment by rcombs** [comment]
> Based on this discussion, I suppose your original intention for the `format` attribute was merely to catch cases when the format string is valid-enough but the arguments don’t match?

Yeah, and it immediately caught a few. This is the most common case these warnings tend to run into (particularly in error handling or other less-commonly-hit code paths that a dev might not run into during testing).

**Comment by TheOneric** [comment]
Since afaik our current (not explicitly documented) compile baseline is “ISO-C99 with a few additional assumptions¹” , setting `-Wpedantic` makes sense to me to actually get warnings in line with our baseline target. I won't insist on it though.
¹: *like eg NULL being equal to all-bits zero, special-treatment for WIndows' deficient `fopen` et al, and some cherry-picked type-width constraints from POSIX iirc*

For format specifiers specifically the currently common non-ISO extensions we want to avoid are:
*(There might be more, eg \*BSD extensions I'm not aware of, but I guess those are less likely to be used by any of us)*
```
 base POSIX:  n$  (specifing argument number)
 base POSIX:  '   (modifier for numeric types to print thousand seps)
POSIX + XSI:  %C  (equal to %lc)
POSIX + XSI:  %S  (equal to %ls)
        GNU:  %m  (strerror(errno); also supported by musl and compile-time option for uClibc)
```

Regardless of adding `-Wpedantic` or future baseline changes, I think the callback should get documentation to state *which* specifiers are used, and which *encoding* the passed data is supposed to be *(UTF-8, but due to potential borked files we can't actually guarantee this, so not a strong promise)*.
Eg, if we decide it should be passed to the libc's methods, we could say: *“uses specifiers from ISO-C99 (may be bumped to a newer ISO-C in the future) and specifiers from `inttypes.h”*, otherwise eg a stronger (current ISO baseline)-specifers only or even ISO-C99-specifiers only promise.

**Comment by astiob** [comment]
Bonus from IRC:

On outdated MinGW-w64 (and IIRC on current suddenly-revived MinGW32), `PRId64` is still `"I64d"` even with `-std=gnu99` or whatever we use.

But glibc has its own `I` **flag,** which enables locale-specific digits (à la Windows).

So GCC with `__attribute__((format(gnu_printf...)))` parses `"%" PRId64`, which expands to `"%I64d"`, as “`%d` padded to 64 characters using localized digits”, and accordingly warns that `int` arguments are expected but `long long` values are passed in.

Of course, the root of the problem is our use of system-specific `PRId64` with system-independent `gnu_printf`, but even so, this particular warning is especially disruptive as it claims there’s a legitimate problem, when there isn’t. If we forced GCC to recognize only standard C formats, it would instead complain that `I` is unrecognized, which is more easily shrugged off by the user.

**Comment by TheOneric** [comment]
> (FWIW I don’t think standard printf formats have been extended since C99.)

This was true (afaik) at the time of writing, but fyi now C23 added the new `b` specifier and `w`, `wf`, `H`, `D` and `DD` length modifiers.
